ignite/pkg/cliui/cliui.go

```diff
@@ -0,0 +1,162 @@
+package cliui
+
+import (
+	"errors"
+	"fmt"
+	"github.com/manifoldco/promptui"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/ignite-hq/cli/ignite/pkg/cliui/cliquiz"
+	"github.com/ignite-hq/cli/ignite/pkg/cliui/clispinner"
+	"github.com/ignite-hq/cli/ignite/pkg/cliui/entrywriter"
+	"github.com/ignite-hq/cli/ignite/pkg/cliui/icons"
+	"github.com/ignite-hq/cli/ignite/pkg/events"
+)
+
+type Session struct {
+	ev       *events.Bus
+	eventsWg *sync.WaitGroup
+
+	spinner          *clispinner.Spinner
+	spinnerWasPaused bool
+
+	in          io.Reader
+	out         io.Writer
+	printLoopWg *sync.WaitGroup
+}
+
+type Option func(s *Session)
+
+func WithOutput(output io.Writer) Option {
+	return func(s *Session) {
+		s.out = output
+	}
+}
+
+func WithInput(input io.Reader) Option {
+	return func(s *Session) {
+		s.in = input
+	}
+}
+
+func New(options ...Option) Session {
+	wg := &sync.WaitGroup{}
+	session := Session{
+		spinner:     clispinner.New(),
+		ev:          events.NewBus(events.WithWaitGroup(wg)),
```

Comment:
```go
WithEvents(events chan events.Event) Option {}
```

I think we should init the event bus only if an events chan given through an Option.

This way caller side will be able to control the channel.

If the event bus isn't set, we don't need to start the `printLoop` goroutine and we don't need to set `Session.ev` with an event bus.

Reply:
I think it is more general practice. I our case I believe we have a channel as part of internal logic. Why could user decide to change buffer size in our case or why could it decide to close it. I think user shouldn't know that it is a channel inside of a Session at all. They should know about Send event function and requirement to call Session.Cleanup()